### PR TITLE
Upgrade to lavalink v4

### PR DESCRIPTION
**Pull request by lavecat**
| -             lavecat Commit : Last Change for lavalink V4                                           - |
| - Clean of code and other chose,                                                                            - |
| - By lavecat ---------------------------------------------------------------------------- |

### CODE REVIEW

File: lavaplay/node_manager.py [hunk 1]
```diff
@@ -171,11 +172,12 @@ async def search_soundcloud(self, query: str) -> t.Optional[t.Union[t.List[Track
             If the track could not be loaded.
         """
         result = await self.rest.load_tracks(f"scsearch:{query}")
-        if result["loadType"] == "NO_MATCHES":
+        res = result["data"]
+        if result["loadType"] == "empty":
```
Comment (HazemMeqdad) [inline lavaplay/node_manager.py:176]: It not work

File: lavaplay/player.py [hunk 2]
```diff
@@ -33,8 +33,8 @@ def add_to_queue(self, tracks: t.List[Track], requester: t.Optional[int] = None)
         """
         Add tracks to queue. use to load a playlist result.
 
-        >>> playlist = lavaplayer.search_youtube("playlist url")
-        >>> lavaplayer.add_to_queue(playlist.tracks)
+        >>> playlist = lavaplay.search_youtube("playlist url")
+        >>> lavaplay.add_to_queue(playlist.tracks)
```
Comment (HazemMeqdad) [inline lavaplay/player.py:37]: Enhance the example

Comment (lavecat) [reply lavaplay/player.py:37]: Ah, I hadn't noticed the change. What do you mean, I didn't understand?

File: lavaplay/node_manager.py (hunk 1 above)
Comment (lavecat) [reply lavaplay/node_manager.py:176]: I don't think it's right that it doesn't work the way it should.   ( https://lavalink.dev/api/rest.html#empty-result-data )

Comment (lavecat) [reply lavaplay/node_manager.py:176]: I just tested it with a bot and it works.

File: lavaplay/ws.py [hunk 3]
```diff
@@ -38,22 +36,18 @@ def __init__(
...
             "Authorization": password,
             "User-Id": str(user_id),
-            "Client-Name": f"Lavaplay.py/{__version__}",
-            "Num-Shards": str(shards_count)
+            "Client-Name": f"Lavaplay.py/{__version__}"
         }
         self._loop = loop or node.loop
         self.emitter: Emitter = node.event_manager
         self.is_connect: bool = False
-        self.resume_key = resume_key or self.node._resume_key or generate_resume_key()
```
Comment (HazemMeqdad) [inline lavaplay/ws.py:47]: The same point, Resume key is feature

File: lavaplay/ws.py [hunk 4]
```diff
@@ -38,22 +36,18 @@ def __init__(
         self._headers = {
             "Authorization": password,
             "User-Id": str(user_id),
-            "Client-Name": f"Lavaplay.py/{__version__}",
-            "Num-Shards": str(shards_count)
```
Comment (HazemMeqdad) [inline lavaplay/ws.py:42]: Should keep this

**Review by HazemMeqdad — CHANGES_REQUESTED** [review]

**Comment by HazemMeqdad** [comment]
@lavecat Sorry, I just review it again and leave some comments. Also could you resolve the conflicts?

File: lavaplay/ws.py [hunk 5]
```diff
@@ -102,7 +96,7 @@ async def callback(self, payload: dict):
             await self.node.rest.update_session(
                 self._session_id,
                 data={
-                    "resumingKey": self.resume_key,
+                    "resuming": False,
```
[the remark this answers is not in the document]
Comment (lavecat) [reply lavaplay/ws.py:99]: resumingKey is delete on v4 https://lavalink.dev/api/rest#update-session

File: lavaplay/ws.py (hunk 3 above)
Comment (lavecat) [reply lavaplay/ws.py:47]: https://lavalink.dev/api/rest#update-session

File: lavaplay/ws.py (hunk 4 above)
Comment (lavecat) [reply lavaplay/ws.py:42]: Okay, I do, but why exactly?

Comment (lavecat) [reply lavaplay/ws.py]: https://github.com/HazemMeqdad/lavaplay.py/pull/49/commits/36b4386b25405997082b3987e9a76db884c20c7c Good

Comment (HazemMeqdad) [reply lavaplay/ws.py]: Because shartds still used

**Comment by HazemMeqdad** [comment]
> @HazemMeqdad This is good?

Good, we are ready to merge this PR.

**Review by HazemMeqdad — COMMENTED** [review]
Can you resolve the conflict.

**Comment by lavecat** [comment]
Good?

**Comment by HazemMeqdad** [comment]
> Good?

I can't merge it

image

**Comment by lavecat** [comment]
It works with the branch master but not v4

**Comment by lavecat** [comment]
@HazemMeqdad Is it possible to merge it with the master branch?

**Comment by HazemMeqdad** [comment]
> @HazemMeqdad Is it possible to merge it with the master branch?

It is okay

**Review by HazemMeqdad — APPROVED** [review]

**Comment by lavecat** [comment]
Yes

**Pull request merged**